rust/src/cfd.rs

```diff
@@ -145,8 +149,37 @@ pub async fn open(order: &Order) -> Result<()> {
     Ok(())
 }
 
-pub async fn settle(taker_amount: u64, maker_amount: u64) -> Result<()> {
-    tracing::info!("Settling CFD with taker amount {taker_amount} and maker amount {maker_amount}");
+pub async fn settle(order: Order, offer: Offer) -> Result<()> {
+    // TODO: this method should not expect the order, but rather the CFD, expecting the order model
+    // here since depending functions are on the order. Note, eventually the order model should
+    // probably be part of the cfd model.
+    let total_margin_sats = Decimal::try_from(order.margin_total().0)? * Decimal::from(100_000_000);
+    let taker_margin_sats = Decimal::try_from(order.margin_taker().0)? * Decimal::from(100_000_000);
+
+    let price = match order.position {
+        Position::Long => offer.bid,
+        Position::Short => offer.ask,
+    };
+
+    let taker_amount_sats =
+        Decimal::try_from(order.calculate_profit(price)?.0)? * Decimal::from(100_000_000);
+    // add profit or substract loss from takers margin. we could calculate the payout here as well,
+    // but this is already implicitly included in the profit.
+    let taker_amount_sats = taker_margin_sats + taker_amount_sats;
+    // TODO: adding this only as a safe guard as this should not happen. I guess it could happen
+    // though until we automatically close on liquidation price.
+    let taker_amount_sats = taker_amount_sats.max(Decimal::ZERO);
+
+    let taker_amount_sats = taker_amount_sats
+        .round_dp_with_strategy(8, rust_decimal::RoundingStrategy::MidpointAwayFromZero);
+
+    let maker_amount_sats = total_margin_sats - taker_amount_sats;
+
+    assert_eq!(total_margin_sats, maker_amount_sats + taker_amount_sats);
+
+    tracing::info!(
+        "Settling CFD with taker amount {taker_amount_sats} and maker amount {maker_amount_sats}"
+    );
```

Comment:
A nice feature of tracing is to add the variables as field names. This makes log analyzis easier as the log message always remains the same and the variables are extra, e.g.

```suggestion
    tracing::info!(taker_amount_sats, maker_amount_sats, 
        "Settling CFD"
    );
```

 in json format this will look something like this: 

```json
{
timestamp: xxx
message: "Settling CFD",
fields: {
  taker_amount_sats: yyy,
  maker_amount_sats: zzz,
}
}
```